Introduce Netty OpenSSL engine

fzakaria opened the pull request:
### Description

Add the ability for MiniSSL to try and create Netty's SSLEngine
using OpenSSL bindings through https://netty.io/wiki/forked-tomcat-native.html

It provides much better performance than Java's default SSL engine
as outlined in https://netty.io/wiki/requirements-for-4.x.html#benefits-of-using-openssl

> Speed: In local testing, we've seen performance improvements of 3x over the JDK. GCM, which is used by the only cipher suite required by the HTTP/2 RFC, is 10-500x faster.

The fallback is kept if the proper Netty libraries are not found on
the classpath.

This pull-request also adds some missing _constants_ & _methods_ that were needed on MIniSSL in order for the tests to pass successfully.

### Benchmark

#### JDK Engine

```
  2 threads and 4 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     3.58ms    2.47ms  87.61ms   98.72%
    Req/Sec   580.07     62.48   666.00     90.30%
  Latency Distribution
     50%    3.31ms
     75%    3.49ms
     90%    3.85ms
     99%    6.44ms
  34553 requests in 30.10s, 6.49GB read
Requests/sec:   1147.91
Transfer/sec:    220.86MB
```

#### OpenSSL engine

```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    14.71ms   27.64ms 149.69ms   83.57%
    Req/Sec     2.54k     1.01k    4.97k    69.00%
  Latency Distribution
     50%  477.00us
     75%   11.29ms
     90%   65.76ms
     99%   98.47ms
  151625 requests in 30.07s, 28.49GB read
Requests/sec:   5042.61
Transfer/sec:      0.95GB
```

### Detailed Information

This is a little information for Ruby-ists that might not be familiar with Java/JRuby.
If you are, feel free to skip ahead to the pull request!

Java is a compiled language; meaning that any _Classes_ used must resolve at the moment of compilation -- the end result is a ".class" file. The classes which were present at compilation must also be present at Runtime if you'd like to resolve them through discovery via the Classpath (LOAD_PATH for Java). If they are not found, you are presented with a "ClassNotFoundException".

This runtime mechanism allows code to be compiled against an API but at runtime swap out the implementation or perhaps take a different course of action if it is not included.

This pull-request takes this approach to try and _upgrade_ to using Netty's OpenSSL if it's discovered (i.e. resolves) on the Classpath; if not it degrades back to using Java's Native SSL engine.

As long as the API is stable, other versions at runtime for Netty could be used aside from the version that was used to compile the JAR.

### Your checklist for this pull request
- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [x] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

fzakaria commented on Rakefile:
```diff
@@ -49,6 +49,7 @@ else
   # Java (JRuby)
   Rake::JavaExtensionTask.new("puma_http11", gemspec) do |ext|
     ext.lib_dir = "lib/puma"
+    ext.classpath = "netty-handler-4.1.47.Final.jar:netty-buffer-4.1.47.Final.jar"
```
This was just added so that compilation through the _JavaExtensionTask_ succeeds.
The documentation should outline which version of Netty it is compiled against.
Luckily, Netty is very aggressive about keeping API compatibility -- so I suspect all 4.X to work.

nateberkopec replied:
Hmmm, any way we can not point directly to an exact version here? Would be better if could just specify netty 4, as you suggested.

fzakaria replied:
Hey @nateberkopec ; not sure how familiar you are with Java. When compiling the JAR for distribution you have to compile the Java files into Class files.
In order to achieve this, the Netty library usage within MiniSSL.java needs to be resolvable at compile time.
At **runtime** however, the dependency can be omitted at which point the JVM throws an exception (ClassNotFoundException).

This classpath library here is not analogous to a Gemfile listing a version range; so there is no equivalent to "~> 4.0"; it has to be an actual JAR to build against.

nateberkopec replied:
Very unfamiliar with Java. I'm also going to have to figure out how this works with our rake-compile and release flow.

fzakaria commented on ext/puma_http11/org/jruby/puma/MiniSSL.java:
```diff
@@ -196,6 +192,61 @@ public IRubyObject initialize(ThreadContext threadContext, IRubyObject miniSSLCo
+                                               long verify_mode,
+                                               String[] ciphers) throws NoSuchAlgorithmException, KeyManagementException {
+    try {
+      io.netty.handler.ssl.ClientAuth clientAuth = io.netty.handler.ssl.ClientAuth.NONE;
```
If Netty libraries are not available this will throw a _ClassNotFoundException_ and default back to the original implementation.

nateberkopec replied:
Does this work? In CI, the build simply fails https://travis-ci.org/github/puma/puma/jobs/661773151#L326

fzakaria replied:
Looks like the JARs were not included in the PR; I will add them.

fzakaria commented on ext/puma_http11/org/jruby/puma/MiniSSL.java:
```diff
@@ -196,6 +192,61 @@ public IRubyObject initialize(ThreadContext threadContext, IRubyObject miniSSLCo
+      if ((verify_mode & 0x1) != 0) {
+        clientAuth = io.netty.handler.ssl.ClientAuth.OPTIONAL;
+      }
+      if ((verify_mode & 0x2) != 0) {
```
https://github.com/netty/netty/blob/2f32e0b8adb63decd9031e26fa5dd4154d93ce97/handler/src/main/java/io/netty/handler/ssl/JdkSslContext.java#L346
Are the associated mappings to verify.

nateberkopec commented:
Thanks for your contribution 🙇 

If any other JRuby users are watching, would love your thoughts as well.

fzakaria commented:
Tagging @headius but not sure if he has the bandwidth;

nateberkopec commented:
Also, do we need to make this configurable? I don't really have an opinion, but just noticing here that we're using Netty _as long as it exists_, which may not be appropriate for a security-related feature.

fzakaria commented on JRUBY.md:
````diff
@@ -0,0 +1,41 @@
+-J-Dpuma.ssl.use-netty=true
+```
+
+Currently, Puma has been compiled against version **4.1.47.Final** of Netty; however it is likely that other
````
Again this is what it is "compiled against"; but Java resolves dependencies at runtime via the classpath.
Customers are free to change the version they run against.

nateberkopec replied:
Is there really no better way in Java to deal with this? "We compiled against X.Y.Z but other versions _may_ work" is not great.

fzakaria replied:
It's actually very similar to C in that regard.

For C, the macro _OPENSSL_VERSION_NUMBER_ defines the library Puma was compiled against and _OpenSSL_version(OPENSSL_VERSION)_ is the runtime call to check the version it was loaded with at runtime.

it's a side effect of dynamic libraries.

fzakaria commented on Rakefile:
```diff
@@ -46,10 +48,29 @@ if !Puma.jruby?
+       https://repo1.maven.org/maven2/io/netty/netty-codec/4.1.47.Final/netty-codec-4.1.47.Final.jar
+       https://repo1.maven.org/maven2/io/netty/netty-buffer/4.1.47.Final/netty-buffer-4.1.47.Final.jar
+       https://repo1.maven.org/maven2/io/netty/netty-handler/4.1.47.Final/netty-handler-4.1.47.Final.jar
+       https://repo1.maven.org/maven2/io/netty/netty-tcnative-boringssl-static/2.0.29.Final/netty-tcnative-boringssl-static-2.0.29.Final.jar).each do |dependency|
```
netty-tcnative-boringssl-static is a really useful dependency. It makes pulling in boringssl (a fork of OpenSSL) extremely easy.

> This is an uber jar that contains the statically linked libraries for all supported netty-tcnative-boringssl-static-{os_arch}. While this tends to be a fairly large jar, it greatly simplifies the getting started experience since an application need not be concerned with obtaining the correct jar for the platform.

nateberkopec replied:
I'm confused on why we have to add it?

fzakaria replied:
I supply this to _RakeTestTask_ so that the Puma SSL tests on JRuby can easily use OpenSSL without any setup needed for the environment.

The alternative is to dynamically link against OpenSSL on the system but that's more fragile than statically compiling in BoringSSL directly.

NB: BoringSSL is just a slimmed down fork of OpenSSL by Google.

fzakaria commented on ext/puma_http11/org/jruby/puma/MiniSSL.java:
```diff
@@ -143,6 +161,7 @@ public IRubyObject initialize(ThreadContext threadContext, IRubyObject miniSSLCo
     KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
     KeyStore ts = KeyStore.getInstance(KeyStore.getDefaultType());
 
+    String alias = miniSSLContext.callMethod(threadContext, "alias").convertToString().asJavaString();
```
Consumers must now provide the _alias_ of the certificate name within the KeyStore.
So that we know which PrivateKey to pull out to setup as the TLS for the server.

```
keytool -list -v -keystore examples/puma/keystore.jks -storepass blahblah                                                    
Keystore type: jks
Keystore provider: SUN

Your keystore contains 1 entry

Alias name: mydomain
Creation date: Aug 23, 2012
Entry type: PrivateKeyEntry
```

NB* I would have preferred to avoid KeyStore entirely and work directly with PEM files; but I find this to be less disruptive.

fzakaria commented on ext/puma_http11/org/jruby/puma/MiniSSL.java:
```diff
@@ -47,6 +65,36 @@ public static void createMiniSSL(Ruby runtime) {
+        ssl.defineConstant("OPENSSL_LIBRARY_VERSION", runtime.newString("Unknown"));
+        ssl.defineConstant("OPENSSL_VERSION", runtime.newString("Unknown"));
+      }
+    }
```
These constants were missing which would actually cause the test file **test_puma_server_ssl.rb** to actually skip itself for JRuby because _DISABLE_SSL_ was true due to the missing constants.

We now have test coverage again for JRuby + SSL !
🍾 🎆 

> rake test TEST=test/test_puma_server_ssl.rb
> Fabulous run in 0.895517s, 12.2834 runs/s, 10.0501 assertions/s.
> 11 runs, 9 assertions, 0 failures, 0 errors, 2 skips

nateberkopec replied:
Thanks for adding this!

nateberkopec commented:
@fzakaria Can you rebase? I moved the JRuby build to Github Actions.

nateberkopec commented on Rakefile:
```diff
@@ -46,10 +48,29 @@ if !Puma.jruby?
+      filename = URI(dependency).path.split('/').last
+      next if File.exists?(filename)
+      response = Net::HTTP.get_response(URI(dependency))
+      open(filename, "wb") { |file|
```
It would be better if we wrote this to the `pkg` directory (which you'll have to make if it doesnt exist ofc).

fzakaria replied:
wrote it to the _tmp/java/puma_http11_ directory

nateberkopec commented on logging.properties:
```diff
@@ -0,0 +1,4 @@
+handlers= java.util.logging.ConsoleHandler
```
Why does this file have to be added?

fzakaria replied:
MiniSSL.java utilizes some logging methods via JUL (Java Utility Logging).
I found it useful for testing to include this properties file so I can turn on more finer log statements as I developed this change.

I describe it in the _JRUBY.md_. It would not be distributed with Puma and is merely a development productivity file.

kares replied:
are you sure `LOGGER.info("Using JDK SSL engine");` won't get distributed?
instead of these you might consider not using JUL and simply using (J)Ruby's `debug/verbose` flag.
or a system property `-Dpuma.ssl.debug=true` to turn on simple (std-out) logging.

nateberkopec commented on test/test_puma_server_ssl.rb:
```diff
@@ -55,6 +57,7 @@ def start_server
     if Puma.jruby?
       ctx.keystore =  File.expand_path "../../examples/puma/keystore.jks", __FILE__
       ctx.keystore_pass = 'blahblah'
+      ctx.alias = 'mydomain'
```
I think we should add something about this in the HISTORY ("`alias must be provided`")

nateberkopec commented:
Thanks for your patience - your explanations are really helping me out @fzakaria!

fzakaria commented:
@nateberkopec thanks for all the great feedback.

1. I've rebased with the latest changes
2. I've reworked the code so there's no need for _alias_ any longer.
3. I've included a benchmark! 🔥  (please see update description)

But the numbers are impressive at least for these 100kB documents.

**OpenSSL** vs **JDK Engine**
477.00us **vs** 3.31ms
151625 requests in 30.07s, 28.49GB read **vs** 34553 requests in 30.10s, 6.49GB read
5042.61 requests/sec **vs** 1147.91 requests/sec
0.95GB Transfer/sec **vs** 220.86MB transfer/sec

olleolleolle commented:
Noted that the GitHub Actions builds for JRuby have the failure described in https://github.com/jruby/jruby-openssl/issues/172 in them:

> Calling SSLContext#min_version= or max_version= results in call to undefined method set_minmax_proto_version

MSP-Greg commented:
Re the `SSLContext#min_version=` issues, see PR https://github.com/puma/puma/pull/2186. Came across this issue in another repo.

JRuby is using much of the MRI ruby code for SSLContext, and the ruby code calls a  private method ([`SSLContext#set_minmax_proto_version`](https://msp-greg.github.io/ruby_trunk/openssl/OpenSSL/SSL/SSLContext.html#set_minmax_proto_version-instance_method)) that is a c method/function, which doesn't exist in JRuby.  That method calls an OpenSSL method that was added in 1.1.0.

Not sure if the fix will work with Netty...

fzakaria commented:
I will kick off the test again with the rebase.

MSP-Greg commented:
@fzakaria

I just looked at BoringSSL's [ssl.h](https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html), and there are differences between it and OpenSSL as to the min/max protocol selection API.  Not sure how that affects JRuby's min/max methods, or the PR I did...

fzakaria commented:
Looking at the remaining failure; I don't think it's related to the change but the test itself is flaky

The test does:
```ruby
system "ruby -rrubygems -Ilib bin/pumactl -F test/shell/t2_conf.rb start &"
sleep 5
system "curl http://localhost:10103/"
```
so it just waits 5 seconds before trying to curl.

However we can see in the output of the GitHub action that it took longer than 5 seconds.
You can see the startup message occurs after the cURL
```
curl: (7) Failed to connect to localhost port 10103: Connection refused
Puma starting in single mode...
* Version 4.3.3 (jruby 9.2.11.0 - ruby 2.5.7), codename: Mysterious Traveller
* Min threads: 0, max threads: 16
* Environment: development
* Listening on tcp://[::]:10103
``

nateberkopec commented:
The integration tests that run with the script file are super flaky. There's an open issue to rewrite those into proper tests. They're relics from before we had a good integration test suite.

nateberkopec commented:
I think for now, we can't merge. Given the discussion on #2188, we should first attempt a stdlib implementation, which then may lead to changes in JRuby itself on SSL optimization. 

We may come back to this after that.

Thanks for all of your hard work and fast responses on this PR 🙏 

I've brought in a similar benchmark to what you wrote to master. Still working on it.
